File: android/app/src/main/kotlin/com/transport_system/ts_admin/data_providers/database/DatabaseManager.kt
package com.transport_system.ts_admin.data_providers.database

import android.content.Context
import com.transport_system.ts_admin.data_providers.database.extensions.decodeGroupInnerConversations
import com.transport_system.ts_admin.data_providers.database.extensions.decodeGroupSettings
import com.transport_system.ts_admin.data_providers.database.extensions.decodeOtoConversation
import com.transport_system.ts_admin.data_providers.database.extensions.getGroupDatabase
import com.transport_system.ts_admin.data_providers.database.extensions.getOtoDatabase
import com.transport_system.ts_admin.data_providers.models.group.GroupConversationModel
import com.transport_system.ts_admin.data_providers.models.group.GroupInnerConversationModel
import com.transport_system.ts_admin.data_providers.models.oto.ConversationModel
import com.transport_system.ts_admin.helpers.AppLogger
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext


class DatabaseManager {

    companion object{
        val instance = DatabaseManager()

        // database names
        const val OTO_CONVERSATIONS_DATABASE_NAME = "conversations_database.db"
          const val GROUP_CONVERSATIONS_DATABASE_NAME = "group_conversations_database.db"

        //
        // table names
        const val OTO_CONVERSATIONS_TABLE_NAME = "conversations"
         const val GROUP_CONVERSATIONS_TABLE_NAME = "group_conversations"

    }

    //
    //
    //function to fetch the specific conversation with ID
    suspend fun fetchOtoConversation(conversationId: Int?, context: Context): ConversationModel? {
        if (conversationId == null) {
            AppLogger.log("nil conversationId passed in fetchOtoConversation")
            return null
        }
        return withContext(Dispatchers.IO){
            val db = getOtoDatabase(context = context) ?: return@withContext null
             try {
                val query = "SELECT * FROM $OTO_CONVERSATIONS_TABLE_NAME WHERE conversation_id = ?;"
                val cursor = db.rawQuery(query, arrayOf(conversationId.toString())).use { cursor ->
                    var conversation: ConversationModel? = null
                    if (cursor.moveToFirst()) {
                        val conversationJsonString = cursor.getString(2)
                        conversation = conversationJsonString?.takeIf { it.isNotEmpty() }
                            ?.let { decodeOtoConversation(it) }
                    }
                    conversation
                }
                cursor
            } catch (e: Exception) {
                AppLogger.log("Error fetching OTO conversation ===> ${e.message}")
                null
            } finally {
                db.close()
            }
        }
    }

    //
    //
    // function to fetch the specific inner conversation from the group with ID
    suspend fun fetchGroupInnerConversation(conversationId: Int?, context: Context): GroupInnerConversationModel? {
        if (conversationId == null) {
            AppLogger.log("nil conversationId passed in fetchGroupInnerConversation")
            return null
        }
        return withContext(Dispatchers.IO){
            val db = getGroupDatabase(context) ?: return@withContext null
             try {
                db.rawQuery("SELECT * FROM $GROUP_CONVERSATIONS_TABLE_NAME", null)?.use { cursor ->
                    var conversation: GroupInnerConversationModel? = null
                    while (cursor.moveToNext()) {
                        try {
                            val jsonString = cursor.getString(2)
                            val conversations = decodeGroupInnerConversations(jsonString)
                            conversation = conversations?.firstOrNull { it.id == conversationId }
                            if (conversation != null) {
                                break
                            }
                        } catch (e: Exception) {
                            AppLogger.log("Error decoding conversation JSON or processing cursor: ${e.message}")
                        }
                    }
                    conversation
                }
            } catch (e: Exception) {
                AppLogger.log("Error during database query execution in fetchGroupInnerConversation: ${e.message}")
                null
            } finally {
                try {
                    db.close()
                } catch (e: Exception) {
                    AppLogger.log("Error closing database in fetchGroupInnerConversation: ${e.message}")
                }
            }
        }
    }

    //
    //
    // function to fetch the specific group contains a conversations ID
    suspend fun fetchGroupDetails(conversationId: Int?, context: Context): GroupConversationModel? {
        if (conversationId == null) {
            AppLogger.log("nil conversationId passed in fetchGroupDetails")
            return null
        }
        return withContext(Dispatchers.IO){
            val db = getGroupDatabase(context = context) ?: return@withContext null
            var group: GroupConversationModel? = null
             try {
                db.rawQuery("SELECT * FROM $GROUP_CONVERSATIONS_TABLE_NAME", null).use { cursor ->
                    // Loop through the cursor rows
                    while (cursor.moveToNext()) {
                        try {
                            val conversationsJsonString = cursor.getString(2)
                            val conversations = decodeGroupInnerConversations(conversationsJsonString)
                            if (conversations != null) {
                                val found = conversations.firstOrNull { it.id == conversationId }
                                if (found != null) {
                                    group = GroupConversationModel(
                                        id = cursor.getInt(0),
                                        name = cursor.getString(1),
                                        conversations = conversations,
                                        conversationsCount = cursor.getInt(4),
                                        unreadCount = cursor.getInt(5),
                                        groupSettings = cursor.getString(3)?.let { groupSettingsJsonString ->
                                            decodeGroupSettings(groupSettingsJsonString)
                                        }
                                    )
                                    break
                                }
                            }
                        } catch (e: Exception) {
                            AppLogger.log("Error processing cursor row: ${e.message}")
                        }
                    }
                    group
                }
            }
            catch (e:Exception){
                AppLogger.log("Error during database query execution in fetchGroupDetails: ${e.message}")
                null
            }
            finally {
                try {
                    db.close()
                } catch (e: Exception) {
                    AppLogger.log("Error closing database in fetchGroupDetails: ${e.message}")
                }
            }
        }
    }

}

File: ios/Runner/Database/DatabaseManager.swift
//
//  DatabaseManager.swift
//  Runner
//
//  Created by Hashim Khan on 16/12/2024.
//

import SQLite3
import Foundation

class DatabaseManager {
    
    static let shared = DatabaseManager()
    
    //
    // database names
    static let OtoConversationsDatbaseName = "conversations_database.db"
    static let GroupConversationsDatbaseName = "group_conversations_database.db"
    
    //
    // table names
    static let OtoConversationsTableName = "conversations"
    static let GroupConversationsTableName = "group_conversations"
    
    
    
    
    
    //
    //
    // function to fetch the specific conversation with ID
    func fetchOtoConversation(conversationId: Int?) throws -> ConversationModel?  {
        
        guard let conversationId = conversationId else {
            print("nil conversationId passed in fetchOtoConversation")
            return nil
        }
        
        guard let dbPath = getOtoConversationsDatabasePath() else {
            print("Empty Database Path")
            return nil
        }
        
        var db: OpaquePointer?
        
        if (sqlite3_open(dbPath, &db) != SQLITE_OK) {
            print("Error opening database")
            return nil
        }
        
        defer {
            sqlite3_close(db)
        }
        
        let query = "SELECT * FROM \(DatabaseManager.OtoConversationsTableName) WHERE conversation_id = ?;"
        var statement: OpaquePointer?
        
        guard sqlite3_prepare_v2(db, query, -1, &statement, nil) == SQLITE_OK else {
            print("Error preparing query: \(String(cString: sqlite3_errmsg(db)))")
            return nil
        }
        
        sqlite3_bind_int64(statement, 1, sqlite3_int64(conversationId))
        
        var conversation: ConversationModel? = nil
        
        if sqlite3_step(statement) == SQLITE_ROW {
            if let conversationJsonString = extractJsonString(from: statement, at: 2) {
                conversation = decodeOtoConversation(from: conversationJsonString)
            }
        }
        
        sqlite3_finalize(statement)
        
        return conversation
    }
    
    
    
    //
    //
    // function to fetch the specific inner conversation from the group with ID
    func fetchGroupInnerConversation(conversationId: Int?) throws -> GroupInnerConversationModel?  {
        
        guard let conversationId = conversationId else {
            print("nil conversationId passed in fetchGroupConversation")
            return nil
        }
        
        guard let dbPath = getGroupConversationsDatabasePath() else {
            print("Empty Database Path")
            return nil
        }
        
        var db: OpaquePointer?
        
        if (sqlite3_open(dbPath, &db) != SQLITE_OK) {
            print("Error opening database")
            return nil
        }
        
        defer {
            sqlite3_close(db)
        }
        
        let query = "SELECT * FROM \(DatabaseManager.GroupConversationsTableName);"
        var statement: OpaquePointer?
        
        guard sqlite3_prepare_v2(db, query, -1, &statement, nil) == SQLITE_OK else {
            print("Error preparing query: \(String(cString: sqlite3_errmsg(db)))")
            return nil
        }
        
        
        var conversation: GroupInnerConversationModel? = nil
        
        while sqlite3_step(statement) == SQLITE_ROW {
            guard let conversationsJsonString = extractJsonString(from: statement, at: 2),
                  let conversations = decodeGroupInnerConversations(from: conversationsJsonString) else {
                continue
            }
            
            if let found = conversations.first(where: { $0.id == conversationId }) {
                conversation = found
                break
            }
        }
        
        sqlite3_finalize(statement)
        
        return conversation
    }
    
    
    
    //
    //
    // function to fetch the specific group contains a conversations ID
    func fetchGroupDetails(conversationId: Int?) throws -> GroupConversationModel?  {
        
        guard let conversationId = conversationId else {
            print("nil conversationId passed in fetchGroupInnerConversation")
            return nil
        }
        
        guard let dbPath = getGroupConversationsDatabasePath() else {
            print("Empty Database Path")
            return nil
        }
        
        var db: OpaquePointer?
        
        if (sqlite3_open(dbPath, &db) != SQLITE_OK) {
            print("Error opening database")
            return nil
        }
        
        defer {
            sqlite3_close(db)
        }
        
        let query = "SELECT * FROM \(DatabaseManager.GroupConversationsTableName);"
        var statement: OpaquePointer?
        
        guard sqlite3_prepare_v2(db, query, -1, &statement, nil) == SQLITE_OK else {
            print("Error preparing query: \(String(cString: sqlite3_errmsg(db)))")
            return nil
        }
        
        
        var group: GroupConversationModel? = nil
        
        while sqlite3_step(statement) == SQLITE_ROW {
            guard let conversationsJsonString = extractJsonString(from: statement, at: 2),
                  let conversations = decodeGroupInnerConversations(from: conversationsJsonString) else {
                continue
            }
            
            if conversations.first(where: { $0.id == conversationId }) != nil {
                
                group = GroupConversationModel()
                group?.id = extractInt(from: statement, at: 0)
                group?.name = extractString(from: statement, at: 1)
                group?.conversations = conversations
                group?.conversationsCount = extractInt(from: statement, at: 4)
                group?.unreadCount = extractInt(from: statement, at: 5)
                if let groupSettingsJsonString = extractJsonString(from: statement, at: 3),
                   let groupSettings = decodeGroupSettings(from: groupSettingsJsonString) {
                    group?.groupSettings = groupSettings
                }
                break
            }
        }
        
        sqlite3_finalize(statement)
        
        return group
    }
    
    
    
    
    
    
}
